examples/applications/cloud-app/src/Program.cs

```diff
@@ -2,109 +2,482 @@
+}
+catch (Exception ex)
+{
+    Console.Error.WriteLine(ex);
+    return 1;
+}
 
+static async Task<int> Main(ProgramArguments args)
+{
+    var jsonSerializerOptions = new JsonSerializerOptions { WriteIndented = true };
+    var utf8 = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+
+    var timeout = new Timeout(TimeSpan.FromSeconds(int.Parse(args.OptTimeout, NumberStyles.None, CultureInfo.InvariantCulture)));
+
+    var mqttFactory = new MqttFactory();
     using var mqttClient = mqttFactory.CreateMqttClient();
 
-    var mqttClientOptions =
-        new MqttClientOptionsBuilder()
-            .WithTcpServer(args is [var server, ..] ? server : "localhost")
-            .WithProtocolVersion(MqttProtocolVersion.V500)
-            .Build();
+    await timeout.ApplyAsync(cancellationToken =>
+    {
+        var options =
+            mqttFactory.CreateClientOptionsBuilder()
+                       .WithTcpServer(args.OptBroker)
+                       .WithProtocolVersion(MqttProtocolVersion.V500)
+                       .Build();
+        return mqttClient.ConnectAsync(options, cancellationToken);
+    });
 
-    await mqttClient.ConnectAsync(mqttClientOptions, CancellationToken.None);
     Console.Error.WriteLine("The MQTT client is connected.");
 
-    mqttClient.ApplicationMessageReceivedAsync += args =>
+    var rpcClient = new ChariottRpcClient(mqttFactory, mqttClient);
+
+    await timeout.ApplyAsync(cancellationToken =>
     {
-        args.Dump();
-        return Task.CompletedTask;
+        var options =
+            mqttFactory.CreateSubscribeOptionsBuilder()
+                       .WithTopicFilter(ChariottRpcClient.ResponseWildcardTopic)
+                       .Build();
+        return mqttClient.SubscribeAsync(options, cancellationToken);
+    });
+
+    var binName = Path.GetFileNameWithoutExtension(Environment.ProcessPath);
+
+    var eventsTopic = string.Join("/", Environment.MachineName,
+                                       binName,
+                                       Environment.ProcessId,
+                                       "events");
+
+    const string eventsFileExtension = ".cjson"; // https://en.wikipedia.org/wiki/JSON_streaming#Concatenated_JSON
+    var eventFilesDirPath = Path.Join(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), $".{binName}", "events");
+    Directory.CreateDirectory(eventFilesDirPath);
+
+    var oldDate = DateTime.Today.AddDays(-30);
+    var oldEventsFiles =
+        from f in new DirectoryInfo(eventFilesDirPath).EnumerateFiles("*" + eventsFileExtension, new EnumerationOptions())
+        where f.CreationTime.Date < oldDate
+        select f;
+
+    foreach (var file in oldEventsFiles)
+        file.Delete();
+
+    var eventsFilePath = Path.Join(eventFilesDirPath, eventsTopic.Replace('/', '=') + eventsFileExtension);
+    var eventsFileReadPosition = 0L;
+    var eventsFileLock = new SemaphoreSlim(1);
```

Comment:
Not strictly needed because the wait handle is never used, but let's err on the safe side in case the implementation changes in the future so committed with ae1dbf29e8a882e23f1058c6a694e82c88133c76.